File: data_manipulation/easy/ads_performance/ads_performance.scala
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.sql.functions._

class AdsPerformanceCalculator {
  def calculate(adsDf: DataFrame): DataFrame = {
    val clickedDf = adsDf
      .groupBy("ad_id")
      .agg(sum(when(col("action") === "Clicked", lit(1)).otherwise(lit(0))).as("clicked"))

    val totalDf = adsDf
      .groupBy("ad_id")
      .agg(sum(when(col("action").isin("Clicked", "Viewed"), lit(1)).otherwise(lit(0))).as("total"))

    clickedDf
      .alias("t1")
      .join(totalDf.alias("t2"), col("t1.ad_id") === col("t2.ad_id"), "inner")
      .select(
        col("t1.ad_id").as("ad_id"),
        coalesce(
          round((col("clicked").cast("double") / nullif(col("total").cast("double"), lit(0.0))) * 100, 2),
          lit(0.0)
        ).as("ctr")
      )
      .orderBy(col("ctr").desc, col("ad_id"))
  }
}

object AdsPerformance {
  def solve(adsDf: DataFrame): DataFrame = {
    val calculator = new AdsPerformanceCalculator()
    calculator.calculate(adsDf)
  }

  def run(spark: SparkSession): DataFrame = {
    val adsDf = spark.table("ads")
    solve(adsDf)
  }
}

File: data_manipulation/easy/ads_performance/ads_performance.py
from pyspark.sql import DataFrame, functions as F


def ads_performance(ads: DataFrame) -> DataFrame:

    result = ads.groupBy("ad_id")\
        .agg(
            F.sum(
                F.when(F.col("action")=="Clicked",1) \
                .otherwise(0)
                ).alias("clicked_count"), 
            F.sum(
                F.when(F.col("action") \
                .isin("Clicked", "Viewed"),1) \
                .otherwise(0)
                ).alias("total")
            ) \
        .withColumn(
            "ctr",
                F.when(F.col("total") == 0, F.lit(0.0)) \
                .otherwise(
                    F.round(
                        (F.col("clicked_count") / F.col("total")) * 100,
                    2))) \
        .select("ad_id", "ctr")

    return result
